operator/slot_ticker/slotticker.go

```diff
@@ -6,83 +6,82 @@ import (
+	slotDuration time.Duration
+	genesisTime  time.Time
 }
 
-// Done should be called to clean up the ticker.
-func (s *SlotTicker) Done() {
-	go func() {
-		s.done <- struct{}{}
-	}()
+func (cfg SlotTickerConfig) SlotDurationSec() time.Duration {
+	return cfg.slotDuration
 }
 
-// NewSlotTicker starts and returns a new SlotTicker instance.
-func NewSlotTicker(genesisTime time.Time, secondsPerSlot uint64) *SlotTicker {
-	if genesisTime.IsZero() {
-		panic("zero genesis time")
-	}
-	ticker := &SlotTicker{
-		c:    make(chan phase0.Slot),
-		done: make(chan struct{}),
-	}
-	ticker.start(genesisTime, secondsPerSlot, time.Since, time.Until, time.After)
-	return ticker
+func (cfg SlotTickerConfig) GetGenesisTime() time.Time {
+	return cfg.genesisTime
 }
 
-func (s *SlotTicker) start(
-	genesisTime time.Time,
-	secondsPerSlot uint64,
-	since, until func(time.Time) time.Duration,
-	after func(time.Duration) <-chan time.Time) {
+type slotTicker struct {
+	timer        *time.Timer
+	slotDuration time.Duration
+	genesisTime  time.Time
+	slot         phase0.Slot
+}
 
-	d := time.Duration(secondsPerSlot) * time.Second
+func NewSlotTicker(cfgProvider ConfigProvider) SlotTicker {
+	genesisTime := cfgProvider.GetGenesisTime()
+	slotDuration := cfgProvider.SlotDurationSec()
 
-	go func() {
-		sinceGenesis := since(genesisTime)
+	now := time.Now()
+	timeSinceGenesis := now.Sub(genesisTime)
 
-		var nextTickTime time.Time
-		var slot phase0.Slot
-		if sinceGenesis < d {
-			// Handle when the current time is before the genesis time.
-			nextTickTime = genesisTime
-			slot = 0
-		} else {
-			nextTick := sinceGenesis.Truncate(d) + d
-			nextTickTime = genesisTime.Add(nextTick)
-			slot = phase0.Slot(nextTick / d)
-		}
+	var initialDelay time.Duration
+	if timeSinceGenesis < 0 {
+		// Genesis time is in the future
+		initialDelay = -timeSinceGenesis // Wait until the genesis time
+	} else {
+		slotsSinceGenesis := timeSinceGenesis / slotDuration
+		nextSlotStartTime := genesisTime.Add((slotsSinceGenesis + 1) * slotDuration)
+		initialDelay = time.Until(nextSlotStartTime)
+	}
 
-		for {
-			waitTime := until(nextTickTime)
-			select {
-			case <-after(waitTime):
-				s.c <- slot
-				slot++
-				nextTickTime = nextTickTime.Add(d)
-			case <-s.done:
-				return
-			}
+	return &slotTicker{
+		timer:        time.NewTimer(initialDelay),
+		slotDuration: slotDuration,
+		genesisTime:  genesisTime,
+		slot:         0,
+	}
+}
+
+func (s *slotTicker) Next() <-chan time.Time {
+	timeSinceGenesis := time.Since(s.genesisTime)
+	if timeSinceGenesis < 0 {
+		return s.timer.C
+	}
+	if !s.timer.Stop() {
+		// try to drain the channel, but don't block if there's no value
+		select {
+		case <-s.timer.C:
+		default:
 		}
-	}()
+	}
+	slotNumber := uint64(timeSinceGenesis / s.slotDuration)
+	nextSlotStartTime := s.genesisTime.Add(time.Duration(slotNumber+1) * s.slotDuration)
+	s.timer.Reset(time.Until(nextSlotStartTime))
+	s.slot = phase0.Slot(slotNumber + 1)
```

Comment:
If `Next` and `Slot` OR `Next` and `Next` are called simultaneously, there might be a risk of read/write or write/write data race here.

Reply:
Well, I think we could keep it as is if we make sure that `Next` and `Slot` are used correctly from the outside, but it would be better to add a comment for it